# HW3 - PyCitySchools Instructions

Your final report should include each of the following:

District Summary

    Create a high level snapshot (in table form) of the district's key metrics, including:
        Total Schools
        Total Students
        Total Budget
        Average Math Score
        Average Reading Score
        % Passing Math (passing requires a score of 70% or greater)
        % Passing Reading (passing requires a score of 70% or greater)
        Overall Passing Rate (Average of the above two)

School Summary

    Create an overview table that summarizes key metrics about each school, including:
        School Name
        School Type
        Total Students
        Total School Budget
        Per School Budget
        Average Math Score
        Average Reading Score
        % Passing Math
        % Passing Reading
        Overall Passing Rate (Average of the above two)

Top Performing Schools (By Passing Rate)

    Create a table that highlights the top 5 performing schools based on Overall Passing Rate. Include:
        School Name
        School Type
        Total Students
        Total School Budget
        Per School Budget
        Average Math Score
        Average Reading Score
        % Passing Math
        % Passing Reading
        Overall Passing Rate (Average of the above two)

Bottom Performing Schools (By Passing Rate)

    Create a table that highlights the bottom 5 performing schools based on Overall Passing Rate. Include all of the same metrics as above.

Math Scores by Grade

    Create a table that lists the average Math Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

Reading Scores by Grade

    Create a table that lists the average Reading Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

Scores by School Spending

    Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
        Average Math Score
        Average Reading Score
        % Passing Math
        % Passing Reading
        Overall Passing Rate (Average of the above two)

Scores by School Size

    Repeat the above breakdown, but this time group schools based on a reasonable approximation of school size (Small, Medium, Large).

Scores by School Type

    Repeat the above breakdown, but this time group schools based on school type (Charter vs. District).

As final considerations:

    Your script must work for both data-sets given.
    You must use the Pandas Library and the Jupyter Notebook.
    You must submit a link to your Jupyter Notebook with the viewable Data Frames.
    You must include an exported markdown version of your Notebook called README.md in your GitHub repository.
    You must include a written description of three observable trends based on the data.
    See Example Solution for a reference on the expected format. Note: the specific fields in that PDF may differ slightly from the instructions. Use the PDF as a guide for the report layout, but refer to the instructions for what fields you should create in each section.

In [1]:
#Import dependencies:
import pandas as pd
import os
from IPython.display import display

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
#Import files
schools_file= os.path.join('raw_data', 'schools_complete.csv')
students_file= os.path.join('raw_data', 'students_complete.csv')
schools_df = pd.read_csv(schools_file)
students_df = pd.read_csv(students_file)
#print(schools_df.head())
#print(schools_df.columns)
#print(students_df.head())
#print(students_df.columns)

In [4]:
#Rename schools_df "name" column to "school" to match other csv file header for school name
schools_df=schools_df.rename(columns= {'name': 'school'})
#print(schools_df.head())
#schools_df.columns
#print(len(schools_df))
#print(len(students_df))

In [5]:
#Merge csv files based on the column 'school'
merge_df = pd.merge(schools_df, students_df, on="school")
#print(merge_df.head())
#print(len(merge_df))

# District Summary

In [6]:
#District Summary#
#Total Schools
total_schools = schools_df['school'].count()

#Total Students
student_count = students_df['name'].count()

#Total Budget
total_budget = schools_df['budget'].sum()

#Average Math Score -NEED TO ROUND IT TO 2 DECIMAL POINTS - DF.ROUND(2)
avg_math = students_df['math_score'].mean()

#Average Reading Score
avg_read = students_df['reading_score'].mean()

#% Passing Math (passing requires a score of 70% or greater)
passing_math= students_df.loc[(students_df["math_score"] >= 70)]
pass_math_count = passing_math["name"].count()
percMathPass = (pass_math_count/student_count)*100

#% Passing Reading (passing requires a score of 70% or greater)
passing_read= students_df.loc[(students_df["reading_score"] >= 70)]
pass_read_count = passing_read["name"].count()
percReadPass = (pass_read_count/student_count)*100

#Overall Passing Rate (Average of the above two)
overall_pass= (percMathPass+percReadPass)/2

#Output of District Summary
district_summary = pd.DataFrame ({"Total Schools": [total_schools], 
                               "Total Students": [student_count],
                               "Total Budget": [total_budget], 
                              "Average Math Score" : [avg_math],
                              "Average Reading Score" : [avg_read],
                              "% Passing Math": [percMathPass],
                               "% Passing Reading": [percReadPass],
                               "% Overall Passing Rate":[overall_pass]})
#print(district_summary)

#Reorganize the columns using double brackets
organized_dist_summary = district_summary[["Total Schools","Total Students","Total Budget","Average Math Score", 
                                 "Average Reading Score", "% Passing Math", "% Passing Reading", 
                                 "% Overall Passing Rate"]]
display(organized_dist_summary)

Total Schools  Total Students  Total Budget  Average Math Score  \
0             15           39170      24649428           78.985371   

   Average Reading Score  % Passing Math  % Passing Reading  \
0               81.87784       74.980853          85.805463   

   % Overall Passing Rate  
0               80.393158

# School Summary

In [7]:
#School Summary# 
#School Name
#School Type
#Total Students
#Total School Budget

school_summary = schools_df
school_summary = school_summary.drop(['School ID'], axis=1) #axis 1 indicates drop that column; default is to drop a row

#Per Student Budget
per_student_budget = school_summary["budget"]/school_summary["size"]
school_summary["Per Student Budget"] = per_student_budget
#print(school_summary)

#Average Math Score
#Average Reading Score
group_school = merge_df.groupby (["school"])
group_school.count()
avg_scores=group_school['math_score','reading_score'].mean().reset_index()
#print(avg_scores)

#Merge avg_scores with school summary on school - MERGE DIDN'T WORK UNTIL I RESET THE INDEX FOR AVG_SCORE. WHY?
school_summary2 = pd.merge(school_summary, avg_scores, on='school')

#Calculate number of students per school
group_school_size= group_school["name"].count()
#print(group_school_size)

#% Passing Math 
school_passing_math = passing_math.groupby(["school"])
sch_pass_math_count = school_passing_math['math_score'].count()

perMathPass_sch = sch_pass_math_count/group_school_size*100
PerMathPass_sch_df= pd.DataFrame(perMathPass_sch)
PerMathPass_sch_df = PerMathPass_sch_df.reset_index()

#% Passing Reading
school_passing_read = passing_read.groupby(["school"])
sch_pass_read_count = school_passing_read['reading_score'].count()

perReadPass_sch = sch_pass_read_count/group_school_size*100
PerReadPass_sch_df = pd.DataFrame(perReadPass_sch)
PerReadPass_sch_df = PerReadPass_sch_df.reset_index()

#%Overall Passing Rate (Average of the above two)
overall_pass_sch = (perMathPass_sch+perReadPass_sch)/2
overall_pass_sch_df = overall_pass_sch.to_frame()
overall_pass_sch_df = overall_pass_sch_df.reset_index()
#print (PerMathPass_sch_df)
#print (PerReadPass_sch_df)
#print(overall_pass_sch_df)

#Merge last three columns to school_summary2 -I WAS ONLY ABLE TO MERGE THESE AFTER RESETTING THE INDEX FOR DFS.WHY?
school_summary3=pd.merge(pd.merge(pd.merge(school_summary2,PerMathPass_sch_df,on='school'),
                                  PerReadPass_sch_df,on='school'),overall_pass_sch_df,on='school')

#Rename column names
#print (school_summary3.columns) # last column heading is an integer 0, not a string '0'.
school_summary_final = school_summary3.rename(columns = {'school' : '',
                                'type' : 'School Type',
                                'size' : 'Total Students',
                                'budget' : 'Total School Budget',
                                'math_score' : 'Average Math Score',
                                'reading_score' : 'Average Reading Score',
                                '0_x' : '% Passing Math',
                                '0_y' : '% Passing Reading',
                                0 : '% Overall Passing Rate'
                                                        })
#Made school the index for the final output
school_summary_final=school_summary_final.set_index("")

#print as table form
display(school_summary_final) 

School Type  Total Students  Total School Budget  \
                                                                         
Huang High School        District            2917              1910635   
Figueroa High School     District            2949              1884411   
Shelton High School       Charter            1761              1056600   
Hernandez High School    District            4635              3022020   
Griffin High School       Charter            1468               917500   
Wilson High School        Charter            2283              1319574   
Cabrera High School       Charter            1858              1081356   
Bailey High School       District            4976              3124928   
Holden High School        Charter             427               248087   
Pena High School          Charter             962               585858   
Wright High School        Charter            1800              1049400   
Rodriguez High School    District            3999              2547363   
Johnson High School      District            4761              3094650   
Ford High School         District            2739              1763916   
Thomas High School        Charter            1635              1043130   

                       Per Student Budget  Average Math Score  \
                                                                
Huang High School                   655.0           76.629414   
Figueroa High School                639.0           76.711767   
Shelton High School                 600.0           83.359455   
Hernandez High School               652.0           77.289752   
Griffin High School                 625.0           83.351499   
Wilson High School                  578.0           83.274201   
Cabrera High School                 582.0           83.061895   
Bailey High School                  628.0           77.048432   
Holden High School                  581.0           83.803279   
Pena High School                    609.0           83.839917   
Wright High School                  583.0           83.682222   
Rodriguez High School               637.0           76.842711   
Johnson High School                 650.0           77.072464   
Ford High School                    644.0           77.102592   
Thomas High School                  638.0           83.418349   

                       Average Reading Score  % Passing Math  \
                                                               
Huang High School                  81.182722       65.683922   
Figueroa High School               81.158020       65.988471   
Shelton High School                83.725724       93.867121   
Hernandez High School              80.934412       66.752967   
Griffin High School                83.816757       93.392371   
Wilson High School                 83.989488       93.867718   
Cabrera High School                83.975780       94.133477   
Bailey High School                 81.033963       66.680064   
Holden High School                 83.814988       92.505855   
Pena High School                   84.044699       94.594595   
Wright High School                 83.955000       93.333333   
Rodriguez High School              80.744686       66.366592   
Johnson High School                80.966394       66.057551   
Ford High School                   80.746258       68.309602   
Thomas High School                 83.848930       93.272171   

                       % Passing Reading  % Overall Passing Rate  
                                                                  
Huang High School              81.316421               73.500171  
Figueroa High School           80.739234               73.363852  
Shelton High School            95.854628               94.860875  
Hernandez High School          80.862999               73.807983  
Griffin High School            97.138965               95.265668  
Wilson High School             96.539641               95.203679  
Cabrera High School            9

# Top Performing Schools (By Overall Passing Rate)

In [8]:
Top5_Schools=school_summary_final
Top5_Schools=Top5_Schools.nlargest(5,'% Overall Passing Rate')
display(Top5_Schools)

School Type  Total Students  Total School Budget  \
                                                                       
Cabrera High School     Charter            1858              1081356   
Thomas High School      Charter            1635              1043130   
Pena High School        Charter             962               585858   
Griffin High School     Charter            1468               917500   
Wilson High School      Charter            2283              1319574   

                     Per Student Budget  Average Math Score  \
                                                              
Cabrera High School               582.0           83.061895   
Thomas High School                638.0           83.418349   
Pena High School                  609.0           83.839917   
Griffin High School               625.0           83.351499   
Wilson High School                578.0           83.274201   

                     Average Reading Score  % Passing Math  % Passing Reading  \
                                                                                
Cabrera High School              83.975780       94.133477          97.039828   
Thomas High School               83.848930       93.272171          97.308869   
Pena High School                 84.044699       94.594595          95.945946   
Griffin High School              83.816757       93.392371          97.138965   
Wilson High School               83.989488       93.867718          96.539641   

                     % Overall Passing Rate  
                                             
Cabrera High School               95.586652  
Thomas High School                95.290520  
Pena High School                  95.270270  
Griffin High School               95.265668  
Wilson High School                95.203679

# Bottom Performing Schools (By Overall Passing Rate)

In [9]:
Bottom5_Schools=school_summary_final
Bottom5_Schools=Bottom5_Schools.nsmallest(5,'% Overall Passing Rate')
display(Bottom5_Schools)

School Type  Total Students  Total School Budget  \
                                                                         
Rodriguez High School    District            3999              2547363   
Figueroa High School     District            2949              1884411   
Huang High School        District            2917              1910635   
Johnson High School      District            4761              3094650   
Ford High School         District            2739              1763916   

                       Per Student Budget  Average Math Score  \
                                                                
Rodriguez High School               637.0           76.842711   
Figueroa High School                639.0           76.711767   
Huang High School                   655.0           76.629414   
Johnson High School                 650.0           77.072464   
Ford High School                    644.0           77.102592   

                       Average Reading Score  % Passing Math  \
                                                               
Rodriguez High School              80.744686       66.366592   
Figueroa High School               81.158020       65.988471   
Huang High School                  81.182722       65.683922   
Johnson High School                80.966394       66.057551   
Ford High School                   80.746258       68.309602   

                       % Passing Reading  % Overall Passing Rate  
                                                                  
Rodriguez High School          80.220055               73.293323  
Figueroa High School           80.739234               73.363852  
Huang High School              81.316421               73.500171  
Johnson High School            81.222432               73.639992  
Ford High School               79.299014               73.804308

# Math and Reading Scores by Grade

In [15]:
Scores_byGrade=merge_df

#Set school as index and select columns of interest
#Scores_byGrade=Scores_byGrade.set_index('school')
Scores_byGrade= Scores_byGrade[["school","grade", "reading_score", "math_score"]]
#print(Scores_byGrade.head())

#Group by school and grade
Grouped_byGrade=Scores_byGrade.groupby(["school", "grade"])

#Output Math Averages by Grade for each school
Math_byGrade=Grouped_byGrade.mean()
del Math_byGrade['reading_score']
New_Math_byGrade=Math_byGrade.reset_index().pivot(index='school', columns='grade')
New_Math_byGrade.columns= New_Math_byGrade.columns.set_levels(['Average Math Score'], level=0)
display(New_Math_byGrade)

#Output Reading Averages by Grade for each school
Reading_byGrade=Grouped_byGrade.mean()
del Reading_byGrade['math_score']
New_Reading_byGrade=Reading_byGrade.reset_index().pivot(index='school', columns='grade')
#print(New_Reading_byGrade.columns)
#print(New_Reading_byGrade.columns.levels[1])
New_Reading_byGrade.columns= New_Reading_byGrade.columns.set_levels(['Average Reading Score '], level=0)
display(New_Reading_byGrade)

#Reorganize Order of grades in both outputs - use loc method to select range and sort columns?
#New_Math_byGrade= New_Math_byGrade([["9th","10th","11th", "12th"]], level=1, axis=1) 
#x=New_Reading_byGrade.sort_index(level=1)
# DOESN'T WORK - multiIndexed
#x1=New_Reading_byGrade.sort_index(level=1, axis=1)      
#print(x1)

Average Math Score                                 
grade                               10th       11th       12th        9th
school                                                                   
Bailey High School             76.996772  77.515588  76.492218  77.083676
Cabrera High School            83.154506  82.765560  83.277487  83.094697
Figueroa High School           76.539974  76.884344  77.151369  76.403037
Ford High School               77.672316  76.918058  76.179963  77.361345
Griffin High School            84.229064  83.842105  83.356164  82.044010
Hernandez High School          77.337408  77.136029  77.186567  77.438495
Holden High School             83.429825  85.000000  82.855422  83.787402
Huang High School              75.908735  76.446602  77.225641  77.027251
Johnson High School            76.691117  77.491653  76.863248  77.187857
Pena High School               83.372000  84.328125  84.121547  83.625455
Rodriguez High School          76.612500  76.395626  77.690748  76.859966
Shelton High School            82.917411  83.383495  83.778976  83.420755
Thomas High School             83.087886  83.498795  83.497041  83.590022
Wilson High School             83.724422  83.195326  83.035794  83.085578
Wright High School             84.010288  83.836782  83.644986  83.264706

Average Reading Score                                  
grade                                   10th       11th       12th        9th
school                                                                       
Bailey High School                 80.907183  80.945643  80.912451  81.303155
Cabrera High School                84.253219  83.788382  84.287958  83.676136
Figueroa High School               81.408912  80.640339  81.384863  81.198598
Ford High School                   81.262712  80.403642  80.662338  80.632653
Griffin High School                83.706897  84.288089  84.013699  83.369193
Hernandez High School              80.660147  81.396140  80.857143  80.866860
Holden High School                 83.324561  83.815534  84.698795  83.677165
Huang High School                  81.512386  81.417476  80.305983  81.290284
Johnson High School                80.773431  80.616027  81.227564  81.260714
Pena High School                   83.612000  84.335938  84.591160  83.807273
Rodriguez High School              80.629808  80.864811  80.376426  80.993127
Shelton High School                83.441964  84.373786  82.781671  84.122642
Thomas High School                 84.254157  83.585542  83.831361  83.728850
Wilson High School                 84.021452  83.764608  84.317673  83.939778
Wright High School                 83.812757  84.156322  84.073171  83.833333

# Scores by School Spending

In [11]:
Scores_bySpending = school_summary_final
Scores_bySpending = Scores_bySpending[["Per Student Budget", "Average Math Score", "Average Reading Score",
                                        "% Passing Math", "% Passing Reading", "% Overall Passing Rate"]]
#print(Scores_bySpending["Per Student Budget"].quantile([0.25, 0.5, 0.75,1]))

#Create bins based on quantiles and create names for each bin
bins = [0, 592, 628, 642, 655]
bin_names= [' < 592', '593-628', '629-642', '643-655']
pd.cut(Scores_bySpending["Per Student Budget"], bins, labels= bin_names)

#Add Spending Ranges per student column - WHAT DOES THE ERROR BELOW MEAN?
Scores_bySpending["Spending Ranges Per Student ($)"] = pd.cut(Scores_bySpending["Per Student Budget"], bins, labels= bin_names)
del Scores_bySpending["Per Student Budget"]

#Perform a groupby based on newly created Spending Ranges column
Grouped_byRange = Scores_bySpending.groupby("Spending Ranges Per Student ($)")
display(Grouped_byRange.mean())

Average Math Score  Average Reading Score  \
Spending Ranges Per Student ($)                                              
 < 592                                    83.455399              83.933814   
593-628                                   81.899826              83.155286   
629-642                                   78.990942              81.917212   
643-655                                   77.023555              80.957446   

                                 % Passing Math  % Passing Reading  \
Spending Ranges Per Student ($)                                      
 < 592                                93.460096          96.610877   
593-628                               87.133538          92.718205   
629-642                               75.209078          86.089386   
643-655                               66.701010          80.675217   

                                 % Overall Passing Rate  
Spending Ranges Per Student ($)                          
 < 592                                        95.035486  
593-628                                       89.925871  
629-642                                       80.649232  
643-655                                       73.688113

# Scores by School Size

In [12]:
Scores_bySchoolSize = school_summary_final
Scores_bySchoolSize = Scores_bySchoolSize[["Total Students", "Average Math Score", "Average Reading Score",
                                        "% Passing Math", "% Passing Reading", "% Overall Passing Rate"]]

#Create bins based on quantiles and create names for each bin
bins2 = [0, 1786, 2925, 4976]
bin_names2= ['  Small (<1786)', ' Medium (1787-2925)', 'Large (2926-4976)']
pd.cut(Scores_bySchoolSize["Total Students"], bins2, labels= bin_names2)

#Add Spending Ranges per student column - WHAT DOES THE ERROR BELOW MEAN?
Scores_bySchoolSize["School Size"] = pd.cut(Scores_bySchoolSize["Total Students"], bins2, labels= bin_names2)
del Scores_bySchoolSize["Total Students"]
#print(Scores_bySchoolSize.columns)

#Perform a groupby based on newly created School Size column
Grouped_bySize = Scores_bySchoolSize.groupby("School Size")
display(Grouped_bySize.mean())

Average Math Score  Average Reading Score  \
School Size                                                      
  Small (<1786)               83.554500              83.850220   
 Medium (1787-2925)           80.750065              82.769850   
Large (2926-4976)             76.993025              80.967495   

                     % Passing Math  % Passing Reading  % Overall Passing Rate  
School Size                                                                     
  Small (<1786)           93.526422          96.500267               95.013345  
 Medium (1787-2925)       83.065610          90.161203               86.613407  
Large (2926-4976)         66.369129          80.995600               73.682364

# Scores by School Type (Charter or District)

In [13]:
Scores_bySchoolType = school_summary_final
Scores_bySchoolType = Scores_bySchoolType[["School Type", "Average Math Score", "Average Reading Score",
                                        "% Passing Math", "% Passing Reading", "% Overall Passing Rate"]]
#print(Scores_bySchoolType)

#Perform a groupby based on School Type column
Grouped_bySchoolType = Scores_bySchoolType.groupby("School Type")
display(Grouped_bySchoolType.mean())

Average Math Score  Average Reading Score  % Passing Math  \
School Type                                                              
Charter               83.473852              83.896421       93.620830   
District              76.956733              80.966636       66.548453   

             % Passing Reading  % Overall Passing Rate  
School Type                                             
Charter              96.586489               95.103660  
District             80.799062               73.673757

# Three observable trends from the data above
1.	District-wise kids score higher, on average, in reading (81.88%) compared to math (74.98%).  
2.	The difference between average math and reading scores is about the same (within a one point difference) for the top 5 performing schools, whereas the difference between average math and reading scores for the bottom 5 performing schools is about 3-4 points. Interestingly the top 5 performing schools are all charter schools, whereas the bottom performing schools are all district schools. The overall passing rate for these top 5 schools is ~95.1%, whereas the passing rate for the bottom 5 schools is 73.7% (see Scores by School Type table). Initially I thought that the difference in performance was related to funding, but that is not the case. The top 5 schools actually have smaller per student budgets ranging from 578-638 dollars, whereas the bottom 5 schools have larger budgets of 637-655 dollars per student. The ‘Scores by School Spending’ table confirms that higher spending per student does not translate to higher overall passing rate, in fact it is the opposite. 
3.	The highest overall passing rate is strongly associated with school size, with students succeeding at higher rates in smaller schools (with <1786 students) compared to larger schools (with 2926-4976 students). In this dataset, school size is a big factor influencing the higher success rate of Charter schools compared to the lower success rate in District schools, with the latter having two to four times as many students as the former.
